### PR TITLE
locale: also save X11 keyboard settings to /etc/vconsole.conf

### LINKED ISSUE #24228: Request for localectl to write the X11 keymap to /etc/vconsole.conf
### Component

other

### Is your feature request related to a problem? Please describe

I hope I am describing this issue correctly, this came up on a draft merge request for Plymouth to make it so it doesn't rely on the kernel mode console for input, but evdev instead.

localectl writes the keymaps to the kernel console in /etc/vconsole.conf, but writes the value for the X11 keymaps in an X11 file, and it seems like it's smart enough to translate the keymap names, As I understand X11 (or user mode) keymap names are not one-to-one with the console keymap names

User mode processes that are not X11, and can't understand X11 config files easily might want a single place to refer to the default configured keymap that libxkbcommon understands.

Having the value saved in  /etc/vconsole.conf by localectl might be good because the file is already known, maybe as an additional key like `XKB_KEYMAP`

### Describe the solution you'd like

It might be hard since now the value will be in two places since one would have to win, (like if someone manually edits the value on vconsole.conf, and not the X11 file, and now the current default one has to be read) but writing the XKB keymap as a key such as `XKB_KEYMAP=foo` to vconsole.conf would allow things that are not Xorg, such as some Wayland servers, or Plymouth. (OK mostly Plymouth) 

Plymouth on distros that use dracut already have /etc/vconsole.conf in their initrd, so amending this file might be the best bet for it

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Closes #24228.
Replaces #25412.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @n3rdopolis.

**Comment by n3rdopolis** [comment]
Thanks! it seems to work cc @halfline

File: src/locale/localed-util.c [hunk 1]
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
         return SYSTEMD_LANGUAGE_FALLBACK_MAP;
 }
 
+static void x11_context_clear(X11Context *xc) {
+        assert(xc);
+
+        xc->layout  = mfree(xc->layout );
+        xc->options = mfree(xc->options);
+        xc->model   = mfree(xc->model  );
+        xc->variant = mfree(xc->variant);
```
Comment (poettering) [inline src/locale/localed-util.c:65]: aligning the `=` makes sense to me, but aligning the `)` goes a little bit too far I think... dunno...

File: src/locale/localed-util.c [hunk 2]
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
...
+}
+
+static bool x11_context_isempty(const X11Context *xc) {
+        assert(xc);
+
+        return
+                isempty(xc->layout ) &&
+                isempty(xc->model  ) &&
+                isempty(xc->variant) &&
```
Comment (poettering) [inline src/locale/localed-util.c:74]: similar

File: src/locale/localed-util.c [hunk 3]
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
...
+
+void x11_context_empty_to_null(X11Context *xc) {
+        assert(xc);
+
+        /* Do not call x11_context_clear() for the passed object. */
+
+        xc->layout  = (char*) empty_to_null(xc->layout);
+        xc->model   = (char*) empty_to_null(xc->model);
+        xc->variant = (char*) empty_to_null(xc->variant);
+        xc->options = (char*) empty_to_null(xc->options);
```
Comment (poettering) [inline src/locale/localed-util.c]: → https://github.com/systemd/systemd/pull/25829

File: src/locale/localed-util.c [hunk 4]
```diff
@@ -421,30 +511,27 @@ int vconsole_convert_to_x11(Context *c) {
...
-                                        return -ENOMEM;
-
-                                modified = true;
-                        }
+                        r = x11_context_copy(xc,
+                                             &(X11Context) {
+                                                     .layout  = (char*) empty_or_dash_to_null(a[1]),
+                                                     .model   = (char*) empty_or_dash_to_null(a[2]),
+                                                     .variant = (char*) empty_or_dash_to_null(a[3]),
+                                                     .options = (char*) empty_or_dash_to_null(a[4]),
```
Comment (poettering) [inline src/locale/localed-util.c]: as above

File: src/locale/localed.c [hunk 5]
```diff
@@ -471,12 +417,37 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
             free_and_strdup(&c->vc_keymap_toggle, keymap_toggle) < 0)
                 return -ENOMEM;
 
+        if (convert) {
+                r = x11_read_data(c, m);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read x11 keyboard layout data: %m");
```
Comment (poettering) [inline src/locale/localed.c]: x11 is usually written X11

File: src/locale/localed.c [hunk 6]
```diff
@@ -471,12 +417,37 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
...
+                        return sd_bus_error_set(error, SD_BUS_ERROR_FAILED, "Failed to read x11 keyboard layout data");
```
Comment (poettering) [inline src/locale/localed.c]: ditto

File: src/locale/localed.c [hunk 7]
```diff
@@ -670,13 +663,13 @@ static int method_set_x11_keyboard(sd_bus_message *m, void *userdata, sd_bus_err
...
-        if (convert) {
-                r = x11_convert_to_vconsole_and_emit(c, m);
-                if (r < 0)
-                        log_error_errno(r, "Failed to convert keymap data: %m");
-        }
+        if (convert)
+                (void) vconsole_reload(sd_bus_message_get_bus(m));
 
         return sd_bus_reply_method_return(m, NULL);
 }
```
Comment (poettering) [inline src/locale/localed.c:661]: afaics this will make some file IO things fatal that previously weren't fatal, no?

I think it would make sense to make only issues with updating our internal idea what settings are in effect fatal. Any changes to the files on disk should probably be handled gracefully: logged about but otherwise ignored.

Comment (poettering) [reply src/locale/localed.c:661]: rationale for the latter: we now write multiple files and we cannot possibly gurantee that either both updates worked or neither. unix fs apis simply have no concept that would allow us to implement that. hence I think we should update what we can, log about what we cannot, but return success as long as our in-memory state was updated.

File: src/locale/localed-util.c [hunk 8]
```diff
@@ -353,6 +373,25 @@ int vconsole_write_data(Context *c) {
...
+
+                r = strv_env_assign(&l, "XKB_MODEL", empty_to_null(xc->model));
+                if (r < 0)
+                        return r;
+
+                r = strv_env_assign(&l, "XKB_VARIANT", empty_to_null(xc->variant));
+                if (r < 0)
+                        return r;
+
+                r = strv_env_assign(&l, "XKB_OPTIONS", empty_to_null(xc->options));
```
Comment (poettering) [inline src/locale/localed-util.c]: one of those days we might want to add strv_env_assign_many() or so with a varargs parm

**Comment by poettering** [comment]
looks pretty good

**Comment by yuwata** [comment]
@poettering Thank you for the review. All requests are addressed. PTAL.

File: src/locale/localed-util.c [hunk 9]
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
...
+}
+
+bool x11_context_is_safe(const X11Context *xc) {
+        assert(xc);
+
+        return
+                (!xc->layout  || string_is_safe(xc->layout) ) &&
+                (!xc->model   || string_is_safe(xc->model)  ) &&
+                (!xc->variant || string_is_safe(xc->variant)) &&
+                (!xc->options || string_is_safe(xc->options));
```
Comment (poettering) [inline src/locale/localed-util.c:96]: still don't like the `)` alignment

File: src/locale/localed-util.c [hunk 10]
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
...
+
+bool x11_context_equal(const X11Context *a, const X11Context *b) {
+        assert(a);
+        assert(b);
+
+        return
+                streq_ptr(a->layout,  b->layout ) &&
+                streq_ptr(a->model,   b->model  ) &&
+                streq_ptr(a->variant, b->variant) &&
+                streq_ptr(a->options, b->options);
```
Comment (poettering) [inline src/locale/localed-util.c:107]: here neither

File: src/locale/localed-util.c [hunk 11]
```diff
@@ -310,17 +402,17 @@ int x11_write_data(Context *c) {
...
 
-        if (!isempty(c->x11_variant))
-                fprintf(f, "        Option \"XkbVariant\" \"%s\"\n", c->x11_variant);
+        if (!isempty(xc->variant))
+                fprintf(f, "        Option \"XkbVariant\" \"%s\"\n", xc->variant);
 
-        if (!isempty(c->x11_options))
-                fprintf(f, "        Option \"XkbOptions\" \"%s\"\n", c->x11_options);
+        if (!isempty(xc->options))
+                fprintf(f, "        Option \"XkbOptions\" \"%s\"\n", xc->options);
```
Comment (poettering) [inline src/locale/localed-util.c:466]: one of these days we should add proper escaping here

File: src/locale/localed.c [hunk 12]
```diff
@@ -471,12 +417,35 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
...
+                }
+
+                /* save the result of conversion to emit changed properties later. */
+                convert = r > 0;
+        }
+
         r = vconsole_write_data(c);
         if (r < 0) {
                 log_error_errno(r, "Failed to set virtual console keymap: %m");
                 return sd_bus_error_set_errnof(error, r, "Failed to set virtual console keymap: %m");
```
Comment (poettering) [inline src/locale/localed.c]: should not be fatal i guess

**Comment by n3rdopolis** [comment]
Seeing this merged would be pretty sweet, as I think it's still a blocker for Plymouth AFAIK

File: src/locale/localed-util.c (hunk 11 above)
Comment (yuwata) [reply src/locale/localed-util.c:466]: Escaping? We check the string with `string_is_safe()`, so I do not think escaping for it.

**Comment by yuwata** [comment]
@poettering Addressed the comments about the spurious alignment. About 'escaping', see the above reply. I hope it is not a blocker of this PR. Upgrading the green label.

@n3rdopolis Sorry for late update.

File: src/locale/localed-util.c [hunk 13]
```diff
@@ -86,6 +164,19 @@ void context_clear(Context *c) {
...
 };
 
+static X11Context *context_get_x11_context(Context *c) {
+        assert(c);
+
+        if (!x11_context_isempty(&c->x11_from_xorg))
+                return &c->x11_from_xorg;
+
+        return NULL;
+}
```
Comment (keszybz) [inline src/locale/localed-util.c]: This seems only suitable for read-only use… The return type should be `const`. (For read-only use it's OK to return NULL even if the object exists. But if the caller were to modify the returned object, it would be strange to have one object in `c` and the other one being modified by the caller.)

File: src/locale/localed-util.c [hunk 14]
```diff
@@ -86,6 +164,19 @@ void context_clear(Context *c) {
...
+
+X11Context *context_get_x11_context_safe(Context *c) {
+        return &ASSERT_PTR(c)->x11_from_xorg;
+}
```
Comment (keszybz) [inline src/locale/localed-util.c]: I think this function is just obfuscating what is happening. All callers assert that `c` is not null, so this `ASSERT_PTR` doesn't gain much. And "safe" is unclear: does it mean that it's safe to call with a NULL argument, which is a common use of "_safe" suffix in our codebase? I'd prefer to drop it and access the field directly.

File: src/locale/localed.c [hunk 15]
```diff
@@ -189,6 +189,7 @@ static int property_get_xkb(
                 sd_bus_error *error) {
 
         Context *c = ASSERT_PTR(userdata);
+        X11Context *xc;
```
Comment (keszybz) [inline src/locale/localed.c]: `const`?

File: src/locale/localed.c [hunk 16]
```diff
@@ -471,12 +417,35 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
             free_and_strdup(&c->vc_keymap_toggle, keymap_toggle) < 0)
                 return -ENOMEM;
 
+        if (convert) {
+                r = x11_read_data(c, m);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read X11 keyboard layout data: %m");
+                        return sd_bus_error_set(error, SD_BUS_ERROR_FAILED, "Failed to read X11 keyboard layout data");
```
Comment (keszybz) [inline src/locale/localed.c]: Why is `%m` not used here?

File: src/locale/localed.c [hunk 17]
```diff
@@ -435,10 +435,8 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
         }
 
         r = vconsole_write_data(c);
-        if (r < 0) {
-                log_error_errno(r, "Failed to set virtual console keymap: %m");
-                return sd_bus_error_set_errnof(error, r, "Failed to set virtual console keymap: %m");
-        }
+        if (r < 0)
+                log_warning_errno(r, "Failed to set virtual console keymap, ignoring: %m");
```
Comment (keszybz) [inline src/locale/localed.c]: Preexisting condition, but "write" would be much better than than "set". The latter sounds like it's about configuring the keymap in the kernel, not writing the config to disk.

File: src/locale/localed.c [hunk 18]
```diff
@@ -646,10 +644,8 @@ static int method_set_x11_keyboard(sd_bus_message *m, void *userdata, sd_bus_err
         }
 
         r = x11_write_data(c);
-        if (r < 0) {
-                log_error_errno(r, "Failed to set X11 keyboard layout: %m");
-                return sd_bus_error_set_errnof(error, r, "Failed to set X11 keyboard layout: %m");
-        }
+        if (r < 0)
+                log_warning_errno(r, "Failed to set X11 keyboard layout, ignoring: %m");
```
Comment (keszybz) [inline src/locale/localed.c]: Here too.

File: src/locale/localed-util.c [hunk 19]
```diff
@@ -167,14 +168,18 @@ void context_clear(Context *c) {
 static X11Context *context_get_x11_context(Context *c) {
         assert(c);
 
+        if (!x11_context_isempty(&c->x11_from_vc))
+                return &c->x11_from_vc;
```
Comment (keszybz) [inline src/locale/localed-util.c:180]: Does the vc context have higher priority?

File: src/locale/localed-util.c [hunk 20]
```diff
@@ -240,6 +250,15 @@ int x11_read_data(Context *c, sd_bus_message *m) {
 
         assert(c);
 
+        r = vconsole_read_data(c, m);
+        if (r < 0)
+                return r;
+
+        if (!x11_context_isempty(&c->x11_from_vc)) {
+                log_debug("XKB settings loaded from vconsole.conf, skipping to read xorg.conf.d/00-keyboard.conf.");
```
Comment (keszybz) [inline src/locale/localed-util.c]: "skipping to read" is not grammatical. Either "not reading …" (sounds better) or "skipping reading".

File: src/locale/localed-util.c [hunk 21]
```diff
@@ -351,44 +505,41 @@ static int read_next_mapping(const char* filename,
...
-        map = systemd_kbd_model_map();
+        assert(c);
 
-        if (isempty(c->vc_keymap)) {
-                modified =
-                        !isempty(c->x11_layout) ||
-                        !isempty(c->x11_model) ||
-                        !isempty(c->x11_variant) ||
-                        !isempty(c->x11_options);
+        xc = context_get_x11_context_safe(c);
```
Comment (keszybz) [inline src/locale/localed-util.c:543]: I find the logic in this function very hard to follow. `context_get_x11_context_safe()` returns one of the contexts depending on their contents, and then we operate on it, even though the function is supposed to convert vconsole to x11. Dunno, maybe add an overview comment of what the logic is?

**Review by keszybz — COMMENTED** [review]
This is mostly questions…

File: src/locale/localed-util.c (hunk 14 above)
Comment (yuwata) [reply src/locale/localed-util.c]: Initially this is introduced as a mostly trivial function. But in a later commit, it will be extended.
Here, `_safe` does not mean the safety about the input, but it always return non-NULL value.

File: src/locale/localed-util.c (hunk 13 above)
Comment (yuwata) [reply src/locale/localed-util.c]: That's also for this function. See the comment below. The function will be extended in a later commit, and cannot make this 'const', as it will be used in `_safe()`.

File: src/locale/localed-util.c (hunk 19 above)
Comment (yuwata) [reply src/locale/localed-util.c:180]: Right, since this PR.

File: src/locale/localed-util.c (hunk 21 above)
Comment (yuwata) [reply src/locale/localed-util.c:543]: Sure.

**Comment by yuwata** [comment]
@keszybz Updated. Hopefully, all comments are addressed or answered. PTAL!

File: src/locale/localed-util.c (hunk 14 above)
Comment (dtardon) [reply src/locale/localed-util.c]: Maybe call it `context_ensure_x11_context`?

**Comment by bluca** [comment]
@yuwata needs a rebase

**Comment by bluca** [comment]
> @yuwata needs a rebase

done now

**Comment by keszybz** [comment]
LGTM.

https://github.com/systemd/systemd/issues/24228#issuecomment-1207882785 has a nice overview of why this is useful.

**Comment by poettering** [comment]
See → #26182, which renames the variables.

File: src/locale/localed-util.c [hunk 22]
```diff
@@ -136,6 +256,17 @@ int x11_read_data(Context *c, sd_bus_message *m) {
...
+        assert(c);
+
+        r = vconsole_read_data(c, m);
+        if (r < 0)
+                return r;
+
+        if (!x11_context_isempty(&c->x11_from_vc)) {
+                log_debug("XKB settings loaded from vconsole.conf, not reading xorg.conf.d/00-keyboard.conf.");
+                return 0;
+        }
```
Comment (YHNdnzj) [inline src/locale/localed-util.c:268]: @poettering I'm thinking if this is really desirable...

IMO we should respect more what is actually applied to X11, i.e `00-keyboard.conf` should have higher priority if that exists.

Comment (YHNdnzj) [reply src/locale/localed-util.c:268]: And if the two get desynchronized, localed should be responsible for overwritting `vconsole.conf` based on `00-keyboard.conf`.
